### PR TITLE
Kanuku/feature/compatible mode

### PR DESCRIPTION
**Pull request by kanuku**

### CODE REVIEW

File: api/src/main/java/org/zalando/nakadi/client/java/enumerator/CompatibilityMode.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        this.mode = mode;
+    }
+
+    @JsonValue
+    public String getCompatibilityMode() {
+        return mode;
+    }
+
+    public static Optional<CompatibilityMode> withName(String code) {
+        for (CompatibilityMode e : CompatibilityMode.values()) {
```
Comment (codechimp) [inline api/src/main/java/org/zalando/nakadi/client/java/enumerator/CompatibilityMode.java]: I would add 2 things here:
- null check (fail early)
- sanitize the code before you conduct the check e.g. code.trim().toLowerCase()
- check `e != null` should not be necessary

File: api/src/main/java/org/zalando/nakadi/client/java/model/EventType.java [hunk 2]
```diff
@@ -35,28 +37,74 @@
...
+     *            Statistics of this EventType used for optimization purposes.
+     *            Internal use of these values might change over time. (TBD:
+     *            measured statistics).
+     * @param compatibilityMode
      *
      */
     @JsonCreator
-    public EventType(@JsonProperty("name") String name, @JsonProperty("owning_application") String owningApplication, @JsonProperty("category") EventTypeCategory category,
+    public EventType(@JsonProperty("name") String name, @JsonProperty("owning_application") String owningApplication,
+            @JsonProperty("category") EventTypeCategory category,
```
Comment (codechimp) [inline api/src/main/java/org/zalando/nakadi/client/java/model/EventType.java]: a Builder would make sense here: the constructor's parameter list is huge

File: client/src/main/scala/org/zalando/nakadi/client/java/JavaClientHandler.scala [hunk 3]
```diff
@@ -79,8 +79,10 @@ class JavaClientHandlerImpl(val connection: Connection, subscriber: Subscription
         }
       case HttpResponse(StatusCodes.NotFound, headers, entity, protocol) =>
         Future.successful(Optional.empty())
-      case HttpResponse(status, headers, entity, protocol) if (status.isFailure()) =>
+      case HttpResponse(status, headers, entity, protocol) if (status.isFailure()) =>{
+        logger.error("Failed with HTTP-status {} and reason: "+status.reason(),status.intValue())
```
Comment (codechimp) [inline client/src/main/scala/org/zalando/nakadi/client/java/JavaClientHandler.scala]:
```
logger.error("Failed with HTTP-status {} and reason: {}", status.intValue(), status.reason())
```

File: client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala [hunk 4]
```diff
@@ -53,7 +53,12 @@ class ClientImpl(connection: Connection, subscriber: SubscriptionHandler, charSe
 
   def getEventType(name: String): Future[Either[ClientError, Option[EventType]]] = {
     logFutureEither(
-      connection.get(URI_EVENT_TYPE_BY_NAME.format(name)).flatMap(in => mapToEither(in)(deserializer(eventTypeTR))))
+      connection.get(URI_EVENT_TYPE_BY_NAME.format(name)).flatMap { in => 
+        logger.info("####")
```
Comment (codechimp) [inline client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala]: please remove these logging statements or change the logging level to DEBUG

File: client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala [hunk 5]
```diff
@@ -155,18 +158,18 @@ class ClientImpl(connection: Connection, subscriber: SubscriptionHandler, charSe
   //#  HELPER METHODS  #
   //####################
 
-  private def logFutureEither[A, T](future: Future[Either[ClientError, T]]): Future[Either[ClientError, T]] = {
+  private def logFutureEither[TA, TB](future: Future[Either[ClientError, TB]]): Future[Either[ClientError, TB]] = {
     future recover {
       case e: Throwable =>
-        val msg = s"An unexpected error occured: ${e.getMessage}"
+        val msg = s"An(1) unexpected error occured: ${e.getMessage}"
```
Comment (codechimp) [inline client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala]: what does "An(1)" mean?

File: client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala [hunk 6]
```diff
@@ -155,18 +158,18 @@ class ClientImpl(connection: Connection, subscriber: SubscriptionHandler, charSe
...
         logger.error(msg)
         Left(ClientError(msg, None))
     }
   }
   private def logFutureOption(future: Future[Option[ClientError]]): Future[Option[ClientError]] = {
     future recover {
       case e: Throwable =>
-        val msg = s"An unexpected error occured: ${e.getMessage}"
+        val msg = s"An(2) unexpected error occured: ${e.getMessage}"
```
Comment (codechimp) [inline client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala]: what does "An(2)" mean?

File: client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala [hunk 7]
```diff
@@ -178,7 +181,7 @@ class ClientImpl(connection: Connection, subscriber: SubscriptionHandler, charSe
       case HttpResponse(status, headers, entity, protocol) if (status.isSuccess()) =>
         try {
           Unmarshal(entity).to[String].map { body =>
-            logger.debug(s"Payload: $body")
+            logger.info(s"Payload: $body")
```
Comment (codechimp) [inline client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala]: - actually, the former logging level was more suitable
- please change to logger.debug("Payload: {}", body)
  -> otherwise, a new string is always constructed even though the logging level was turned of

File: client/src/main/scala/org/zalando/nakadi/client/scala/Connection.scala [hunk 8]
```diff
@@ -156,14 +162,19 @@ sealed class ConnectionImpl(val host: String,
   def executeCall(request: HttpRequest): Future[HttpResponse] = {
     logger.debug("executingCall {}", request.method)
     val response: Future[HttpResponse] = Source.single(request).via(requestFlow).runWith(Sink.head)(materializer())
+    response onComplete {
+      case Failure(t) => 
+        logger.error(s"Failed to execute method:${request.method} URL: ${request._2} Error: {}", t.getMessage)
```
Comment (codechimp) [inline client/src/main/scala/org/zalando/nakadi/client/scala/Connection.scala]: why do you mix {} and direct string manipulation (${...})?

File: client/src/main/scala/org/zalando/nakadi/client/utils/ClientBuilder.scala [hunk 9]
```diff
@@ -7,26 +7,33 @@ import org.zalando.nakadi.client.scala.ClientImpl
...
   def apply(host: String = null,
-            port: Int = DEFAULT_PORT,
+            port: Int = DEFAULT_PORT.toInt,
             tokenProvider: Option[() => String] = None,
             securedConnection: Boolean = true,
             verifySSlCertificate: Boolean = true) =
     new ClientBuilder(host, port, tokenProvider, securedConnection, verifySSlCertificate)
 
-  private val DEFAULT_PORT = 443
+  private val DEFAULT_PORT = "443"
```
Comment (codechimp) [inline client/src/main/scala/org/zalando/nakadi/client/utils/ClientBuilder.scala]: why did you change the type from Int to String? If I see it correctly, you always call DEFAULT_PORT.toInt anyway

File: it/src/main/scala/CleanupTestEvents.scala [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  deleteAllEventsCreatedByClientIntegrationTests()
+
+  def deleteAllEventsCreatedByClientIntegrationTests() = {
+    val eventNames = List("SimpleEvent", "ClientIntegrationTest", "PaymentCreatedDataChangeEventTest")
+    val client = ClientFactory.buildScalaClient()
+    val eventsOnNakadi = Await.result(client.getEventTypes(), 25.second)
+    val filteredEvents = eventsOnNakadi match {
+      case Right(Some(events)) if !events.isEmpty=>
+        val result = events.filter(p=> eventNames.exists(s =>p.name.startsWith(s)) && shouldBeRemoved(p.owningApplication) ||shouldBeRemoved(p.owningApplication) )
+        println("######################### Number of filtered events to be deleted: "+result.size)
```
Comment (codechimp) [inline it/src/main/scala/CleanupTestEvents.scala]: please remove or replace with logging. otherwise, we cannot turn of these messages.

File: api/src/main/java/org/zalando/nakadi/client/java/enumerator/CompatibilityMode.java (hunk 1 above)
Comment (kanuku) [inline api/src/main/java/org/zalando/nakadi/client/java/enumerator/CompatibilityMode.java]: * Point 2 is actually the name of the Enum itself (in CAPITAL), which is handy for when you persist something to the database.

File: client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala (hunk 5 above)
Comment (kanuku) [inline client/src/main/scala/org/zalando/nakadi/client/scala/ClientImpl.scala]: Because the Error message was displayed in different functionalities, this was hard to find out where the error actually occurred...

File: client/src/main/scala/org/zalando/nakadi/client/utils/ClientBuilder.scala (hunk 9 above)
Comment (kanuku) [inline client/src/main/scala/org/zalando/nakadi/client/utils/ClientBuilder.scala]: In the default constructor we use the String...

File: it/src/main/scala/CleanupTestEvents.scala (hunk 10 above)
Comment (kanuku) [inline it/src/main/scala/CleanupTestEvents.scala]: This is a "command-line" tool and does not work with logging and was specially made for this a single purpose: Delete created events.

File: client/src/main/scala/org/zalando/nakadi/client/scala/Connection.scala (hunk 8 above)
Comment (kanuku) [inline client/src/main/scala/org/zalando/nakadi/client/scala/Connection.scala]: Because it is  :point_right: possible :point_left:? :face_with_head_bandage:

File: api/src/main/java/org/zalando/nakadi/client/java/enumerator/EnumUtil.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
+package org.zalando.nakadi.client.java.enumerator;
+
+import java.util.Optional;
+
+public class EnumUtil {
```
Comment (codechimp) [inline api/src/main/java/org/zalando/nakadi/client/java/enumerator/EnumUtil.java]: +1
